File: EdDyA II/StacksQueuesDequeues/Dequeue.cpp
#include <iostream>
using namespace std;
template<class t>
class Dequeue{
  public: 
    class Node{
      public: 
        Node *prev, *next;
        t data;
	Node(t data){
      	  this -> data = data;
      	  prev = next = NULL;
      	  cout << "\tNode(" << data << ")" << endl;
      	}
      	~Node(){
      	  cout << "\t~Node(" << data << ")" << endl;
      	}
    };
    Node *first, *last;
    Dequeue(){
      first = last = NULL;
      cout << "\tDequeue()" << endl;
    }
    ~Dequeue(){
      Node *a = first, *b;
      while( a ){
	b = a -> next;
	delete a;
	a = b;
      }
      cout << "\t~Dequeue()" << endl;
    }
    void push_front(t data){
      Node *n = new Node(data);
      if( !first ) last = n;
      else n -> next = first, first -> prev = n;
      first = n;
    }
    void push_back(t data){
      Node *n = new Node(data);
      if( !last ) first = n;
      else n -> prev = last, last -> next = n;
      last = n;
    }
    t pop_front(){
      if( !first ) return (t)NULL;
      t a = first -> data;
      Node *b = first -> next;
      if( first != last ) first -> next -> prev = NULL;
      else last = NULL;
      delete first;
      first = b;
      return a;
    }
    t pop_back(){
      if( !last ) return (t)NULL;
      t a = last -> data;
      Node *b = last -> prev;
      if( first != last ) last -> prev -> next = NULL;
      else first = NULL;
      delete last;
      last = b;
      return a;
    }
    t front(){
      if( !first ) return (t)NULL;
      return first -> data;
    }
    t back(){
      if( !last ) return (t)NULL;
      return last -> data;
    }
    bool isEmpty(){
      return !first;
    }
};
int main(){
  Dequeue<int> d;
  cout << (d.isEmpty()? "Vacia": "Llena") << endl;
  for(int i = 0; i < 10; i++)
    d.push_back(i);
  cout << (d.isEmpty()? "Vacia": "Llena") << endl;
  for(int i = 0; i < 10; i++)
    cout << d.back();
  cout << (d.isEmpty()? "Vacia": "Llena") << endl;
  for(int i = 0; i < 10; i++)
    cout << d.pop_back() << endl;
  cout << (d.isEmpty()? "Vacia": "Llena") << endl;
  for(int i = 0; i < 10; i++)
    d.push_front(i);
  cout << (d.isEmpty()? "Vacia": "Llena") << endl;
  for(int i = 0; i < 10; i++)
    cout << d.front();
  cout << (d.isEmpty()? "Vacia": "Llena") << endl;
  for(int i = 0; i < 10; i++)
    cout << d.pop_front() << endl;
  cout << (d.isEmpty()? "Vacia": "Llena") << endl;
  return 0;
}

File: EdDyA II/StacksQueuesDequeues/Dequeue.py
from __future__ import print_function
class Node:
  def __init__(self, data):
    self.prev = self.next = None
    self.data = data
    print("\tNode(" + str(data) + ")")
class Dequeue:
  def __init__(self): self.first = self.last = None
  def push_front(self, data):
    n = Node(data)
    if not self.first: last = n
    n.next = self.first
    self.first = n
  def push_back(self, data):
    n = Node(data)
    if not self.last: self.first = n
    n.prev = self.last
    self.last = n
  def pop_front(self):
    if not self.first: return
    b = self.first.data
    self.first = self.first.next
    if self.first: self.first.prev = None
    else: self.last = None
    return b
  def pop_back(self):
    if not self.last: return
    b = self.last.data
    self.last = self.last.prev
    if self.last: self.last.next = None
    else: self.first = None
    return b
  def front(self):
    if not self.first: return
    return self.first.data
  def back(self):
    if not self.last: return
    return self.last.data
  def isEmpty(self):
    return False if self.first else True

print("DEQUEUEEEEEEE")

q = Dequeue()
for i in range(1,5):
    q.push_back(i)
for i in range(5,10):
    q.push_front(i)
for i in range(3):
    print(q.front(), end = " ") 
for i in range(3):
    print(q.back(),end = ' ') 
print(q.pop_front())
print(q.pop_back())
for i in range(3):
    print(q.front(),end = ' ') 
for i in range(3):
    print(q.back(),end = ' ') 



for i in range(15):
    print(q.pop_front(),end = ' ') 
    
for i in range(15,19):
    q.push_front(i)
print(q)

while not q.isEmpty():
    print(q.pop_back(),end=' ')#d = Dequeue()

#print(d.isEmpty())
#for i in range(10):
#  d.push_back(i)
#print(d.isEmpty())
#for i in range(10):
#  print(d.back())
#print(d.isEmpty())
#for i in range(10):
#  print(d.pop_back())
#print(d.isEmpty())
#print(d.isEmpty())
#for i in range(10):
#  d.push_front(i)
#print(d.isEmpty())
#for i in range(10):
#  print(d.front())
#print(d.isEmpty())
#for i in range(10):
#  print(d.pop_front())
#print(d.isEmpty())
